src/pde_models.jl

```diff
@@ -0,0 +1,230 @@
+                u[b1,TRIANG[e1,2]]^2*(EDGE[e1,3,1]^2 + EDGE[e1,3,2]^2) +
+                u[b1,TRIANG[e1,3]]^2*(EDGE[e1,1,1]^2 + EDGE[e1,1,2]^2) +
+                2*u[b1,TRIANG[e1,1]]*u[b1,TRIANG[e1,2]]*(EDGE[e1,2,1]*EDGE[e1,3,1] + EDGE[e1,2,2]*EDGE[e1,3,2]) +
+                2*u[b1,TRIANG[e1,1]]*u[b1,TRIANG[e1,3]]*(EDGE[e1,2,1]*EDGE[e1,1,1] + EDGE[e1,2,2]*EDGE[e1,1,2]) +
+                2*u[b1,TRIANG[e1,2]]*u[b1,TRIANG[e1,3]]*(EDGE[e1,1,1]*EDGE[e1,3,1] + EDGE[e1,1,2]*EDGE[e1,3,2])
+            )
+        )
+    end
+    for b1 in 1:dom.BREAK+2
+        energy[b1] = sum(integral[b1, e1] for e1 in 1:dom.ELEM)
+    end
+end
+
+function _initial_position!(problem::PDEProblem, d::PDEDiscretizationDomain, niter)
+    ubar = zeros(d.NODES)
+    u0 = zeros(d.BREAK+2, d.NODES)
+    energy0 = zeros(d.BREAK+2)
+    integral0 = zeros(d.BREAK+2, d.ELEM)
+
+    # Calculate an ending point on the other side of the barrier
+    for i in 1:niter
+        for b1 in 1:d.BREAK+2, n in 1:d.NODES
+            u0[b1, n] = (1 - (b1-1)/(d.BREAK+1))*d.US[n] + ((b1-1)/(d.BREAK+1))*d.UE[n]
+        end
+        _update_values!(integral0, energy0, u0, problem, d)
+        if energy0[end] < 0
+            break
+        end
+        d.UE .*= 2.0
+    end
+
+    # Backtrack to the barrier to get a better representation
+    ubar .= d.US
+    for i in 1:niter
+        for b1 in 1:d.BREAK+2, n in 1:d.NODES
+            u0[b1, n] = (1 - (b1-1)/(d.BREAK+1))*ubar[n] + ((b1-1)/(d.BREAK+1))*d.UE[n]
+        end
+        _update_values!(integral0, energy0, u0, problem, d)
+        ebar = maximum(energy0[b1] for b1 in 1:d.BREAK+2 if energy0[b1] < 0.0)
+        bbar = minimum(b1 for b1 in 1:d.BREAK+2 if energy0[b1] == ebar)
+        ubar .= u0[bbar-1, :]
+        d.UE .= u0[bbar, :]
+    end
+
+    for b1 in 1:d.BREAK+2, n in 1:d.NODES
+        u0[b1, n] = (1 - (b1-1)/(d.BREAK+1))*d.US[n] + ((b1-1)/(d.BREAK+1))*d.UE[n]
+    end
+    _update_values!(integral0, energy0, u0, problem, d)
+    z0 = maximum(energy0[b1] for b1 in 1:d.BREAK+2)
+    return (
+        u=u0, energy=energy0, integral=integral0, z=z0,
+    )
+end
+
+function _transition_state_model(problem, dom::PDEDiscretizationDomain; T = Float64, backend = nothing, kwargs...)
+    a, b, c, d, p = problem.a, problem.b, problem.c, problem.d, problem.p
+    x0 = _initial_position!(problem, dom, 10)
+    _proto_model()
+
+    ALPHA = 2.0
+    H = ALPHA / (dom.BREAK+1) * sqrt(sum((dom.US[n] - dom.UE[n])^2 for n in 1:dom.NODES))
+
+    # Build optimization problem
+    core = ExaModels.ExaCore(T; backend= backend)
+
+    u = ExaModels.variable(core, 1:dom.BREAK+2, 1:dom.NODES; start=x0.u)
+    integral = ExaModels.variable(core, 1:dom.BREAK+2, 1:dom.ELEM)
+    z = ExaModels.variable(core, 1; start=x0.z)
+
+    ExaModels.objective(core, z[1])
+
+    ExaModels.constraint(
+        core,
+        sum(integral[b1+1, e1] for e1 in 1:n) - z[1]
+        for b1 in 2:dom.ELEMS+1;
+        lcon=-Inf,
+        ucon=0.0,
+    )
```

Comment:
These constraints can benefit significantly from `constraint!` if `n` is large.

Consider reformulating this as 
```julia
c = ExaModels.constraint(
    core, 
    - z[1] for b1 in 2:dom.ELEMS+1;
    lcon=-Inf,
    ucon=0.0,
) 
ExaModels.constraint!(
    core,
    b1 -1 => integral[b1+1, e1] for b1 in 2:dom.ELEMS+1
)
```